Code Actions doc page (#4157)
scalameta/metals: docs/codeactions/code-actions.md

title: Code Actions
---

Whenever you are stuck or unsure, please open an issue or [ask us on
Discord](https://discord.gg/DwTc8xbNDd) or [on our Matrix
bridge](https://matrix.to/#/#scalameta:metals-contributors). This project
follows [Scalameta's contribution
guidelines](https://github.com/scalameta/scalameta/blob/master/CONTRIBUTING.md)
and the [Scala CoC](https://scala-lang.org/conduct/).

Code Actions are there to provide refactoring or rewriting of your current code in order to enhance its readability or save you some time.

The common way of using them, is to click on the yellow light bulb appearing on your code, near the cursor and then, to pick the desired action to peform on the code.

Here is a list of the code actions supported by Metals. Some of these code actions support much more complex use cases, which you would eventually discover through your code, by clicking on the magic 💡, as it appears, to get your surprise!

## Convert To Named Arguments

This code action adds the name to parameters passed to a method call or a class instant.

[Convert To Named Arguments Code Action](https://github.com/scalameta/metals/docs/codeactions/gifs/ConvertToNamedArguments.gif)

> ## Create Companion Object

Reviewer: Let's add a short explanation even if it seems obvious

---
id: codeactions
title: Code Actions
---

Code Actions are used to provide automatic refactoring or rewriting of your current code in order to enhance its readability or save users' time.

In Visual Studio Code, to use them you can click on the yellow light bulb appearing on your code, near the cursor and then, to pick the desired action to perform on the code. Alternatively, you can use the `editor.action.quickFix` shortcut.

For other editors this should also be available via specified shortcut.

Here is a list of the code actions supported by Metals. Some of these code actions support much more complex use cases, which you would eventually discover through your code, by clicking on the magic 💡, as it appears, to get your surprise!

## Convert To Named Arguments

This code action adds the name to parameters passed to a method call or a class constructor.

> [Convert To Named Arguments Code Action](https://github.com/scalameta/metals/docs/codeactions/gifs/ConvertToNamedArguments.gif)

Reviewer 2: ```suggestion
![convert to named arguments code action](./gifs/converttonamedarguments.gif)
```

to show the gif in the webpage instead of just a link to gif

## Create Companion Object

If you have a `trait` or a `class` definition, you can invoke this code action to create its companion object.

![Create Companion Object Code Action](./gifs/CreateCompanionObject.gif)

## Create New Symbol

Used for creating a definition for a symbol which is used in your code, but not defined anywhere.

![Create New Symbol Code Action](./gifs/CreateNewSymbol.gif)

## Extract Member

Extract a class, trait, or object definition to a new file.

![Extract Member Code Action](./gifs/ExtractMember.gif)

## Rename File

Renames a file to match that of the prominent type defined in it.

![Rename File](./gifs/RenameFile.gif)

### Extract Value

If you think your code would look nicer, if the calcuation of a value is extracted and assined to a new `val`; this code action would come to your aid.

> ![Extract Value](./gifs/ExtractValue.gif)

Author: ooops, yeah. Thank you so much for taking the time to try this 👍

---
id: codeactions
title: Code Actions
---

Code Actions are used to provide automatic refactoring or rewriting of your current code in order to enhance its readability or save users' time.

In Visual Studio Code, to use them you can click on the yellow light bulb appearing on your code, near the cursor and then, to pick the desired action to perform on the code. Alternatively, you can use the `editor.action.quickFix` shortcut.

For other editors this should also be available via specified shortcut.

> Here is a list of the code actions supported by Metals. Some of these code actions support much more complex use cases, which you would eventually discover through your code, by clicking on the magic 💡, as it appears, to get your surprise!

Reviewer 3: ```suggestion
Here is a list of the code actions supported by Metals.
```
I would omit or rephrase the last sentence here :smile_cat:
This part adds some ambiguity: what means `much more complex use cases`, `you would eventually discover` - when, where?

Author: well, the current gifs are only covering the simplest cases for getting the message across. So let's say the use case of covering partial functions in map/flatMap is not shown, so the user would eventually discover this more complex one through clicking the light bulb.

Reviewer 3: > So let's say the use case of covering partial functions in map/flatMap is not shown

From user viewpoint there is no difference. I think, people would be more surprised if it doesn't work in some cases rather than it support some particular tree inside an expression.

> so the user would eventually discover 

Yep, sometimes actions are discovered by a user but I think it's a rare occasion.

I think, in general, interaction with actions comes not from light-bulb suggestion.
User know which actions Metals supports and calls them to avoid manual rewriting.
So, you want to specify a type of your val - you know that there is an action. The same with for-comp conversion and others. Usually you know in advance that some particular code should have a rewrite action.

Author: ok, no problem.

## Convert To Named Arguments

This code action adds the name to parameters passed to a method call or a class constructor.

![Convert To Named Arguments Code Action](./gifs/ConvertToNamedArguments.gif)

## Create Companion Object

If you have a `trait` or a `class` definition, you can invoke this code action to create its companion object.

![Create Companion Object Code Action](./gifs/CreateCompanionObject.gif)

## Create New Symbol

Used for creating a definition for a symbol which is used in your code, but not defined anywhere.

![Create New Symbol Code Action](./gifs/CreateNewSymbol.gif)

## Extract Member

Extract a class, trait, or object definition to a new file.

![Extract Member Code Action](./gifs/ExtractMember.gif)

## Rename File

Renames a file to match that of the prominent type defined in it.

![Rename File](./gifs/RenameFile.gif)

> ### Extract Value

Reviewer 2: ```suggestion
## Extract Value
```

If you think your code would look nicer, if the calcuation of a value is extracted and assined to a new `val`; this code action would come to your aid.

![Extract Value](./gifs/ExtractValue.gif)

> ### map, flatMap, withFilter to For Yield

Reviewer 2: ```suggestion
## map, flatMap, withFilter to For Yield
```

It converts a chain of `map`, `flatMap`, `filter` and `filterNot` methods into a for comprehension.

![To For Comprehension](./gifs/FlatMapToForComprehension.gif)

## Implement Abstract Members of the Parent Type

Upon inherting from a type, you also have to implement its abstract members. But manually looking them all up and copying their signature is time consuming, isn't it? You can just use this code action instead.

![Implement Abstract Members](./gifs/ImplementAbstractMembers.gif)

> ### Import Missing Symbol 

Reviewer 2: ```suggestion
## Import Missing Symbol 
```

Because, of course, you often do not remember the package where it is located, without using this code action. right?

![Import Missing Symbol](./gifs/ImportMissingSymbol.gif)

## Insert Inferred Type

> This code action tries to infer the type of variables according to the context, and insert them for you.

Reviewer 3: ```suggestion
This code action infers the type of variables according to the context, and insert them for you.
```

---
id: codeactions
title: Code Actions
---

Whenever you are stuck or unsure, please open an issue or [ask us on
Discord](https://discord.gg/DwTc8xbNDd) or [on our Matrix
bridge](https://matrix.to/#/#scalameta:metals-contributors). This project
follows [Scalameta's contribution
guidelines](https://github.com/scalameta/scalameta/blob/master/CONTRIBUTING.md)
and the [Scala CoC](https://scala-lang.org/conduct/).

Code Actions are there to provide refactoring or rewriting of your current code in order to enhance its readability or save you some time.

> The common way of using them, is to click on the yellow light bulb appearing on your code, near the cursor and then, to pick the desired action to peform on the code.

Reviewer: ```suggestion
In Visual Studio Code, to use them you can click on the yellow light bulb appearing on your code, near the cursor and then, to pick the desired action to perform on the code. Alternatively, you can use the `editor.action.quickFix` shortcut.

For other editors this should also be available via specified shortcut.
```

Here is a list of the code actions supported by Metals. Some of these code actions support much more complex use cases, which you would eventually discover through your code, by clicking on the magic 💡, as it appears, to get your surprise!

## Convert To Named Arguments

This code action adds the name to parameters passed to a method call or a class instant.

[Convert To Named Arguments Code Action](https://github.com/scalameta/metals/docs/codeactions/gifs/ConvertToNamedArguments.gif)

## Create Companion Object

> If you have a `trait` or a `class` definition, you can invoke this code action for creating its companion object.

Reviewer: ```suggestion
If you have a `trait` or a `class` definition, you can invoke this code action to create its companion object.
```

## Create New Symbol

This is for creating a definition for a symbol which is used in your code, but not defined anywhere.

[Create New Symbol Code Action](https://github.com/scalameta/metals/docs/codeactions/gifs/CreateNewSymbol.gif)

## Extract Member

Extract a class, trait, or object definition to a new file.

[Extract Member Code Action](https://github.com/scalameta/metals/docs/codeactions/gifs/ExtractMember.gif)

## Rename File

Renames a file to match that of the prominent type defined in it.

[Rename File](https://github.com/scalameta/metals/docs/codeactions/gifs/RenameFile.gif)

### Extract Value

If you think your code would look nicer, if the calcuation of a value is extracted and assined to a new `val`; this code action would come to your aid.
[Extract Value](https://github.com/scalameta/metals/docs/codeactions/gifs/ExtractValue.gif)

### map, flatMap, withFilter to For Yield

It converts a chain of `map`, `flatMap`, `filter` and `filterNot` methods into a for comprehension

[To For Comprehension](https://github.com/scalameta/metals/docs/codeactions/gifs/FlatMapToForComprehension.gif)

## Implement Abstract Members of the Parent Type

> Upon inherting from a type, you also have to implement its abstract members. But manuallylooking them all up and copying their signature is time consuming, isn't it? You can just use this code action instead.

Reviewer: ```suggestion
Upon inherting from a type, you also have to implement its abstract members. But manually looking them all up and copying their signature is time consuming, isn't it? You can just use this code action instead.
```

Author: The informal descriptions are nowadays very popular in technical manuals and blog posts. They entertain and engage the user as they go through the article. So they increase the chance of the article for getting read to the end.

Author: But people are different. However you write it, there will still be people who won't be ok with it, won't there?

[Implement Abstract Members](https://github.com/scalameta/metals/docs/codeactions/gifs/ImplementAbstractMembers.gif)

### Import Missing Symbol 

Because, of course, you often do not remember the package where it is located, without using this code action. right?

[Import Missing Symbol](https://github.com/scalameta/metals/docs/codeactions/gifs/ImportMissingSymbol.gif)

## Insert Inferred Type

This one, tries to infer the type of variables according to the context, and insert them for you.

[Insert Inferred Type](https://github.com/scalameta/metals/docs/codeactions/gifs/InsertInferredType.gif)

## Organize Imports

> Yes, just have all the clutter, as well as the unnecessary ones removed and tidied up.

Reviewer: This is a bit unclear, it would also be good to mention that we need a special compiler option for removing unused imports and that in some editors it's not available on the "lgiht bulb"
